clang/test/SemaCXX/constructor.cpp

```diff
@@ -96,3 +96,13 @@ namespace PR38286 {
   template<typename> struct C; // expected-note {{non-type declaration found}}
   template<typename T> C<T>::~C() {} // expected-error {{identifier 'C' after '~' in destructor name does not name a type}}
 }
+
+namespace GH121706 {
+struct S {
+  *S();  // expected-error {{invalid constructor declaration}}
```

Comment:
As pointed out above, we need a lot more tests. The bug report shows several other cases this applies to and they should all be enumerated in the tests.